packages/tocco-ui/src/Table/Table.js

```diff
@@ -0,0 +1,221 @@
+import PropTypes from 'prop-types'
+import {FormattedMessage} from 'react-intl'
+
+import StyledTable, {
+  PaginationContainer,
+  StretchingTableContainer, StyledDnD, StyledFullRow, StyledFullRowProgress, StyledTableBody,
+  StyledTableHead, StyledTableHeaderCell, StyledTableRow,
+  StyledTableWrapper
+} from './StyledTable'
+import SortingState from './SortingState'
+import ResizingController from './ResizingController'
+import {LoadingSpinner, Pagination, Typography} from '../'
+import useResize from './useResize'
+import useDnD from './useDnD'
+import useSelection from './selection/useSelection'
+import StaticCell from './StaticCell'
+import {selectionStylePropType} from './selection/selectionStyles'
+import {getSelectionCell} from './selection/selectionColumnEnhancer'
+import {columnPropType, dataPropType, keyPropType} from './propTypes'
+
+const InProgressRow = () => <StyledTableRow>
+  <StyledFullRowProgress>
+    <LoadingSpinner size="20"/>
+    <Typography.P>
+      <FormattedMessage id="client.component.table.dataLoading"/>
+    </Typography.P>
+  </StyledFullRowProgress>
+</StyledTableRow>
+
+const NoDataRow = () => <StyledTableRow>
+  <StyledFullRow>
+    <Typography.Span>
+      <FormattedMessage id="client.component.table.noData"/>
+    </Typography.Span>
+  </StyledFullRow>
+</StyledTableRow>
+
+const ThContent = ({column, data}) =>
+  column.HeaderRenderer
+    ? <column.HeaderRenderer column={column} data={data} />
+    : <div dangerouslySetInnerHTML={{__html: column.label}}/> || null
+
+ThContent.propTypes = {
+  data: dataPropType,
+  column: columnPropType
+}
+
+const Table = props => {
+  const [columns, setColumns] = useState(props.columns)
+
+  const tableEl = useRef(null)
+
+  const resizeCallback = (columnId, width) => {
+    setColumns([...columns.map(c => c.id === columnId ? {
+      ...c,
+      width
+    } : c)])
+  }
+
+  const {isSelected, selectionChange}
+    = useSelection(props.selection, props.data.map(e => e.__key), props.onSelectionChange)
+
+  useEffect(() => {
+    const selectionColumn = getSelectionCell(props.selectionStyle, props.columns, isSelected, selectionChange)
+    setColumns([...(selectionColumn ? [selectionColumn] : []), ...props.columns])
+  }, [props.columns, props.selection])
+
+  const {resizingColumn, startResize} = useResize(tableEl, resizeCallback)
+
+  const {dndEvents, dndState} = useDnD(props.onColumnPositionChange, columns)
+
+  const thOnClick = column => e => {
+    if (!resizingColumn && column.sorting.sortable) {
+      props.onSortingChange(column.id, e.shiftKey)
+    }
+  }
+
+  const trOnClick = entity => e => {
+    if (e.shiftKey) {
+      selectionChange(entity.__key, true, true)
+    } else if (e.metaKey || e.ctrlKey) {
+      selectionChange(entity.__key)
+    } else {
+      props.onRowClick(entity.__key)
+    }
+  }
+
+  const showPagination = (props.paginationInfo.totalCount - props.paginationInfo.recordsPerPage) > 0
+  return <StyledTableWrapper>
```

Comment:
maybe this could be broken down into some smaller components to increase readability and decrease complexity (e.g. Header, Body and Row)